libopenage/gamestate/player.cpp

```diff
@@ -111,5 +114,46 @@ void Player::initialise_unit_types() {
 	}
 }
 
+void Player::active_unit_added(Unit *unit) {
+	// check if unit is actually active
+	if (unit->has_attribute(attr_type::building) && unit->get_attribute<attr_type::building>().completed < 1.0f) {
+		return;
+	}
+
+	// TODO handle here building dependencies
+	// TODO handle here on create unit triggers
+
+	// population
+	if (unit->has_attribute(attr_type::population)) {
+		auto popul = unit->get_attribute<attr_type::population>().population;
+		if (popul < 0) {
+			population.add_capacity(-popul);
+		}
+		else if (popul > 0) {
+			population.add_population(popul);
+		}
+	}
+}
+
+void Player::active_unit_removed(Unit *unit) {
+	// check if unit is actually active
+	if (unit->has_attribute(attr_type::building) && unit->get_attribute<attr_type::building>().completed < 1.0f) {
+		return;
+	}
+
+	// TODO handle here building dependencies
+	// TODO handle here on death unit triggers
+
+	// population
+	if (unit->has_attribute(attr_type::population)) {
+		auto popul = unit->get_attribute<attr_type::population>().population;
+		if (popul < 0) {
+			population.add_capacity(popul);
+		}
+		else if (popul > 0) {
+			population.add_population(-popul);
```

Comment:
i'd also implement that as separate functions for freeing and aquiring a slot, e.g. `demand_population` or something like that.